### PR TITLE
Facets review

### PR DESCRIPTION
**Pull request by CBenghi**
Working progress revision of the schema and documentation.

### CODE REVIEW

File: Development/ids.xsd [hunk 1]
```diff
@@ -229,12 +243,30 @@
...
-			<xs:enumeration value="IFCRELASSIGNSTOGROUP"/>
-			<xs:enumeration value="IFCRELCONTAINEDINSPATIALSTRUCTURE"/>
-			<xs:enumeration value="IFCRELNESTS"/>
-			<xs:enumeration value="IFCRELVOIDSELEMENT"/>
-			<xs:enumeration value="IFCRELFILLSELEMENT"/>
+			<xs:enumeration value="IFCRELAGGREGATES"/>  
+			<xs:enumeration value="IFCRELASSIGNSTOGROUP"/> 
+			<xs:enumeration value="IFCRELCONTAINEDINSPATIALSTRUCTURE"/> 
+			<xs:enumeration value="IFCRELNESTS"/> 
+			<xs:enumeration value="IFCRELVOIDSELEMENT IFCRELFILLSELEMENT"/>
```
Comment (andyward) [inline Development/ids.xsd:250]: Assuming this is a typo @CBenghi ? Should be 2 separate enumeration values?

Comment (CBenghi) [reply Development/ids.xsd:250]: It's actually not. It was discussed in the call last Friday.
The idea is to transverse both those relations as if they were one.
Because just one of them does not get you from the window to the wall (the first step is voids).

It would perhaps be possible in the future to expand from this to a list of relations to consider.

File: Development/IDS_random_example.ids [hunk 2]
```diff
@@ -99,21 +99,21 @@
...
-										<Label xmlns="" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">Sixty minutes</Label>
+										<xs:Label xmlns="" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">Sixty minutes</xs:Label>
 									</xs:documentation>
 								</xs:annotation>
 							</xs:enumeration>
 							<xs:enumeration value="90">
 								<xs:annotation>
 									<xs:documentation>
-										<Label xmlns="" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">Ninty minutes</Label>
+										<xs:Label xmlns="" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">Ninty minutes</xs:Label>
```
Comment (pjanck) [inline Development/IDS_random_example.ids:116]:
```suggestion
										<xs:Label xmlns="" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">Ninety minutes</xs:Label>
```

File: Development/ids.xsd [hunk 3]
```diff
@@ -229,12 +243,30 @@
...
+			<xs:enumeration value="Prohibited"/>
+		</xs:restriction>
+	</xs:simpleType>
+	<xs:simpleType name="conditionalCardinality">
+		<xs:restriction base="xs:string">
+			<xs:enumeration value="Required"/>
+			<xs:enumeration value="Prohibited"/>
+			<xs:enumeration value="Optional"/>
 		</xs:restriction>
 	</xs:simpleType>
```
Comment (pjanck) [inline Development/ids.xsd:270]: Propose to use small letters, similarly to `unbounded` from `xs:occurs`.

**Review by pjanck — COMMENTED** [review]
I haven't looked through the documentation changes, since the text seems very WIP today.

File: Development/ids.xsd (hunk 1 above)
Comment (andyward) [reply Development/ids.xsd:250]: OK, that makes sense. I guess the alternative is to decouple from the explicit IFC relationships and rely on a map ( which will be needed any if just for this special case)

**Comment by CBenghi** [comment]
> I haven't looked through the documentation changes, since the text seems very WIP today.

Indeed that is more of a set of working notes that will need to be distributed in the rest of the documentation.
PRs are welcome for that.

File: Documentation/facet-configurations.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    -> IfcElementType.ElementType
+    -> .USERDEFINED.
+  1.2 else Type's PredefinedType (if not null)
+    -> Type's - PredefinedType
+2. the value of the direct attribute PredefinedType
+  2.1 if the attribute PredefinedType == .USERDEFINED. then the value of the attribute IfcElement/ObjectType is used
+  2.2 else entity's PredefinedType (if not null)
+3. TODO document the same for ElementType and ProcessType
+
+TODO: check if the validation service of IFC prevents entity level types to be defined when type's type is specified.
```
Comment (andyward) [inline Documentation/facet-configurations.md:40]: The rationale for this change to the way entity facet's PredefinedType behaviour makes total sense. However it may be worth considering the implications for implementors (both during spec creation and model validation). Particularly for 'early-bound' IFC implementations I can see challenges when traversing from an Element (generally of a known type) to any defining ElementType's PredefinedType enum attribute. The issue being you don't know what Type you'll get at runtime, which provides the enum Type.

Take as an example: an IDS specifier can presumably define an entity facet as:
**Name**: IFCFLOWTERMINAL
**SubType**: BATH  _(the enum from IfcSanitaryTerminalTypeEnum)_

(Obviously in IFC4 onward you might be able to use **Name** = IfcSanitaryTerminal, but not in 2x3)

So two issues:
1. The design interface (and IDS validator) would need to prompt/allow all appropriate enum values for a FlowTerminal -  from all potential DefiningTypes
2. In order to query that PredefinedType value on the ElementType at runtime you've some casting and coercion to do, and a runtime performance hit.

All doable, but it would be a lot more efficient and easier to implement if there were some means to specify the Type as part of the entity. It would also be more expressive: you could select all sanitary equipment in IFC2x3.

I guess the question is: are we considering these kinds of complex uses of PredefinedType/SubType for IDS 1.0? Entity Facets seem most often used on _applicability_ rather than _requirements_ so it's important these kinds of query predicates can be employed consistently across implementations or we'll be testing different sets of applicable items.

**Comment by NickNisbet** [comment]
I need to require that an IfcElement MUST have an IfcElementType.

  *   Is requiring an attribute ‘PredefinedType’ with pattern restriction ‘.*’ the way to do this? (preferred)
  *   Is requiring an entity ‘subType’ with pattern restriction ‘.*’ the way to do this?

Regards,

Nick.


Anhänge! Es sei denn, Sie kennen den Absender und wissen, dass der Inhalt sicher ist.

@andyward commented on this pull request.

________________________________

In Documentation/facet-configurations.md<https://github.com/buildingSMART/IDS/pull/240#discussion_r1479531193>:

> +1. If a type is defined via the IfcRelDefinesByType relation

+  the PredefinedType of the type overrides the entity's PredefinedType (which should be empty according to IFC documentation)

+  1.1 if Type's PredefinedType is .USERDEFINED.

+    -> IfcElementType.ElementType

+    -> .USERDEFINED.

+  1.2 else Type's PredefinedType (if not null)

+    -> Type's - PredefinedType

+2. the value of the direct attribute PredefinedType

+  2.1 if the attribute PredefinedType == .USERDEFINED. then the value of the attribute IfcElement/ObjectType is used

+  2.2 else entity's PredefinedType (if not null)

+3. TODO document the same for ElementType and ProcessType

+

+TODO: check if the validation service of IFC prevents entity level types to be defined when type's type is specified.

The rationale for this change to the way entity facet's PredefinedType behaviour makes total sense. However it may be worth considering the implications for implementors (both during spec creation and model validation). Particularly for 'early-bound' IFC implementations I can see challenges when traversing from an Element (generally of a known type) to any defining ElementType's PredefinedType enum attribute. The issue being you don't know what Type you'll get at runtime, which provides the enum Type.

Take as an example: an IDS specifier can presumably define an entity facet as:
Name: IFCFLOWTERMINAL
SubType: BATH (the enum from IfcSanitaryTerminalTypeEnum)

(Obviously in IFC4 onward you might be able to use Name = IfcSanitaryTerminal, but not in 2x3)

So two issues:

  1.  The design interface (and IDS validator) would need to prompt/allow all appropriate enum values for a FlowTerminal - from all potential DefiningTypes
  2.  In order to query that PredefinedType value on the ElementType at runtime you've some casting and coercion to do, and a runtime performance hit.

All doable, but it would be a lot more efficient and easier to implement if there were some means to specify the Type as part of the entity. It would also be more expressive: you could select all sanitary equipment in IFC2x3.

I guess the question is: are we considering these kinds of complex uses of PredefinedType/SubType for IDS 1.0? Entity Facets seem most often used on applicability rather than requirements so it's important these kinds of query predicates can be employed consistently across implementations or we'll be testing different sets of applicable items.

**Comment by andyward** [comment]
> I need to require that an IfcElement MUST have an IfcElementType. 
> * Is requiring an attribute ‘PredefinedType’ with pattern restriction '.*' the way to do this? (preferred) 
> * Is requiring an entity ‘subType’ with pattern restriction ‘.*’ the way to do this? 
>
> Regards, Nick. 

Nick, I'm not sure either would work reliably - any validator would presumably only flag up issues where there was a PredefinedType attribute value on the ElementType but not on the Element. So as a requirement it might work for my IfcFlowTerminal=>IsTypedBy=>IfcSanitaryTerminalType example above (any value must come from the Type). And it would work for say _IfcBeam_ and _IfcBeamType_ in IFC2x3 **but  not** in  later schemas since IfcBeam.PredefinedType got introduced in Ifc4.  And wouldn't work in any schema for IfcPile since PredefinedType existed on both Element and Type since IFC2x3.

Probably needs a separate issue if there's not one already...

Andy

**Comment by MatthiasWeise** [comment]
> > I need to require that an IfcElement MUST have an IfcElementType.
> > 
> > * Is requiring an attribute ‘PredefinedType’ with pattern restriction '.*' the way to do this? (preferred)
> > * Is requiring an entity ‘subType’ with pattern restriction ‘.*’ the way to do this?
> > 
> > Regards, Nick.
> 
> Nick, I'm not sure either would work reliably - any validator would presumably only flag up issues where there was a PredefinedType attribute value on the ElementType but not on the Element. So as a requirement it might work for my IfcFlowTerminal=>IsTypedBy=>IfcSanitaryTerminalType example above (any value must come from the Type). And it would work for say _IfcBeam_ and _IfcBeamType_ in IFC2x3 **but not** in later schemas since IfcBeam.PredefinedType got introduced in Ifc4. And wouldn't work in any schema for IfcPile since PredefinedType existed on both Element and Type since IFC2x3.
> 
> Probably needs a separate issue if there's not one already...
> 
> Andy

May check this issue (related to IFC2x3 types): https://github.com/buildingSMART/IDS/issues/116

**Comment by gverduci** [comment]
Hi Claudio,
the extension of the attributeType contains the attribute group "xs:occurs" and the new attribute cardinality (line 156). 
Maybe it is a mistake

**Comment by gverduci** [comment]
The entity name in the restriction test case "fail-an_enumeration_matches_case_sensitively_1_3.ids" is in pascal case format
```
<applicability minOccurs="0" maxOccurs="unbounded">
  <entity>
    <name>
      <simpleValue>IfcWall</simpleValue>
    </name>
 </entity>
</applicability>
```

**Comment by CBenghi** [comment]
> Hi Claudio, the extension of the attributeType contains the attribute group "xs:occurs" and the new attribute cardinality (line 156). Maybe it is a mistake

Hi @gverduci, thanks. I think I spotted that too while revising the schema. It should be solved now.

**Comment by gverduci** [comment]
Hi Claudio,
the file "classification/fail-an_optional_classification_value_fails_if_no_match.ids" contains a title and a specification name with a different meaning than the file name.

This file contains an optional requirement with system and value, and the related ifc file contains an IFCWALL without classifications.

For the optional case with system and value, the facet-configuration.md file says:
"OPTIONAL
SYSTEM/VALUE:
entity -> if there is a classification system for the entity to be tested, its value must match"

Is it correct that this test fails?

**Comment by gverduci** [comment]
the file "material/fail-an_optional_material_fails_if_no_value_matches.ids" contains a title and a specification name with a different meaning than the file name

**Comment by CBenghi** [comment]
Hi @gverduci,

with regards to your comments:

> classification/fail-an_optional_classification_value_fails_if_no_match.ids
> Is it correct that this test fails?

I think the origina IFC had a problem, it attached the classification to the project rather than the wall, it should now be ok.
It should fail, because the classification is an empty string and should not match the expected value.

> the file "material/fail-an_optional_material_fails_if_no_value_matches.ids" contains a title and a specification name with a different meaning than the file name

I've changed the code so that we are notified of the title mismatches on testcase regeneration, see 9d37660.

Thanks for spotting these.

**Comment by andyward** [comment]
Hi @CBenghi, when regenerating the testcases (which I assume closes out #192 ?) were you able to fix the invalid IFC for `pass-occurrences_override_the_type_classification_per_system_1_3.ids`

There's a whole bunch of open issues for this one bad test as every implementor seems to hit it: e.g. #108, #194, #225

There's a dead simple fix: change the type of one of the entities so it doesn't get affected by prior test data.

https://github.com/buildingSMART/IDS/pull/226#issuecomment-1850018702

File: Development/IDS_Aedes_example.ids [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
+<ids:ids xmlns:ids="http://standards.buildingsmart.org/IDS" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://standards.buildingsmart.org/IDS http://standards.buildingsmart.org/IDS/0.9.7/ids.xsd">
     <ids:info>
         <ids:title>AedesUVIP-ILS Window use-case example</ids:title>
         <ids:version>1.0</ids:version>
     </ids:info>
     <ids:specifications>
-        <ids:specification ifcVersion="IFC4" name="Beglazing" description="Glazing in a window" minOccurs="0" maxOccurs="unbounded">
-            <ids:applicability>
+        <ids:specification ifcVersion="IFC4" name="Beglazing" description="Glazing in a window" >
+            <ids:applicability minOccurs="0" maxOccurs="unbounded">
```
Comment (vi55i) [inline Development/IDS_Aedes_example.ids:9]: I have never understood what `minOccurs` and `maxOccurs` mean in an `applicability `statement.

Comment (MatthiasWeise) [reply Development/IDS_Aedes_example.ids:9]: There was a long debate about the meaning of min/maxOccurs settings in `specification`. The decision from an engineering point of view was to ask for a specific number of instances that can be selected by `applicability `where ALL instances must fulfill specified requirements. This is the reason to move from `specification `to `applicability`, because this cardinality restriction is purely for number of instances that fulfill the `applicability `statement.

In contrast to that, we did not see a use case asking for a specific number of instances that are selected by applicability AND fulfill specified requirements.
 
Anyhow, most generic EIRs will use min=0 and max=unbounded, meaning that ALL instances fulfilling the applicability settings should fulfill our requirements. There are only few examples where other settings make sense and most of them seem to be addressed either in the IFC schema (1 instance of IfcProject) or implementer agreements (1 instance of IfcSite/IfcBuilding). Other examples like "I need 10 office rooms in my building" are more project specific.

File: Documentation/units.md
Comment (MatthiasWeise) [inline Documentation/units.md:1]: Looks good and seems to be the complete list!
If we want to focus on measurements that should be converted, I think we can reduce the list (essentially all measurements without a link to SI units). 

No need for conversion or unclear conversion process:
* IFCCONTEXTDEPENDENTMEASURE
* IFCCOUNTMEASURE
* IFCDESCRIPTIVEMEASURE
* IFCNUMERICMEASURE
* IFCPHMEASURE

Other questionable measurements are:
* IFCNORMALISEDRATIOMEASURE
* IFCPOSITIVERATIOMEASURE
* IFCRATIOMEASURE
* IFCMONETARYMEASURE (no fixed conversion rate)

Comment (CBenghi) [reply Documentation/units.md:1]: I agree, I've forgot that I left the full list.
I'll improve the file document.

Comment (CBenghi) [reply Documentation/units.md:1]: Done.
Why do you think there's unclear conversion for ph? 
I don't think it's going to need any conversion, and be in the range {0.0 <= SELF <= 14.0}
But there are other such cases in the documentation for completeness.

I think we will have to add a testcases suite for the entire set of measures.

Comment (MatthiasWeise) [reply Documentation/units.md:1]: >Why do you think there's unclear conversion for ph? I don't think it's going to need any conversion, and be in the range {0.0 <= SELF <= 14.0} But there are other such cases in the documentation for completeness.
> 
> I think we will have to add a testcases suite for the entire set of measures.

That was my point, everything what does not need a conversion does not have to be included in the list. If always between 0 and 14, we can remove it.

**Comment by CBenghi** [comment]
Hi @andyward, 
I think I fixed the issue with that IFC file, but I was not following the relevant issues, so I hope I did it right.
I have not implemented the updated verification logic for a while, so your help is much welcome if you have a list of open testcase issues.
Also, now that IDS generation is automated I would like to do the same for IFC, your advice much welcome there too, if you have any.
I'm thinking to produce a number of IFC files that can match the requirements (e.g. combinatorial variation of relevant schemas and element solutions). Think only of the many options that we have when it comes to materials!

**Comment by andyward** [comment]
> Hi @andyward,
I think I fixed the issue with that IFC file, but I was not following the relevant issues, so I hope I did it right.
I have not implemented the updated verification logic for a while, so your help is much welcome if you have a list of open testcase issues.
Also, now that IDS generation is automated I would like to do the same for IFC, your advice much welcome there too, if you have any.
I'm thinking to produce a number of IFC files that can match the requirements (e.g. combinatorial variation of relevant schemas and element solutions). Think only of the many options that we have when it comes to materials!

Sure, I'll see if we can integrate the tests from this PR. Last I checked we pass all the current testcases except 1-2 or esoteric edge cases (ie6 tolerance and Predefined Properties are the only 2 I'm aware of) - so suggest we get it all (regression) tested on 0.9.7, and then can look at additional tests. Is there an XIDS update with the latest IDS schema?

Should be able to help with the IFC generation. That said I think Dion's testcases cover most of the material cases already. Multi-schema would be good to test. I also think Applicability needs some test cases (Currently the tests focus on Requirement). There's a whole bunch of scenarios with restrictions on applicability where it would be good to have a standard set of test cases for.

File: Documentation/units.md
Comment (MatthiasWeise) [reply Documentation/units.md:1]: I was wondering if we should differentiate between three categories:
1) values that may need a conversion before comparison (everything that is defined by SI-units)
2) values that do not need a conversion as there is only one way to represent the data (ph, count, numeric and ratio) 
3) values that shall not be used for comparison as there is no standardised way for representing them (monetary, context dependent, descriptive)  

To be safe, we can exclude category 3 from value checks (except for existance/not existance).
